### ISSUE TITLE
First Attempt at a Windows (Code::Blocks and MinGW) Compile

### ISSUE THREAD

**Issue by msbeck** [daid/EmptyEpsilon#556]
I'm trying to get things compiled on windows in order to make a Windows build. I'm coming up against a few things. I had a lot of other errors, but sorted though those. I've made it as far as building the **debug** exe and running it, but it seems like the DLLs aren't working as expected.

![bin directory of dlls](https://user-images.githubusercontent.com/2982395/44622848-ee2fe380-a886-11e8-9269-fecc5182b6cf.PNG)
All my DLLS

![entry point not found](https://user-images.githubusercontent.com/2982395/44622850-f38d2e00-a886-11e8-8624-33bda63c5bb3.PNG)
![entry point not found 2](https://user-images.githubusercontent.com/2982395/44622851-f556f180-a886-11e8-8384-7b40f7b91796.PNG)
![entry point not found 3](https://user-images.githubusercontent.com/2982395/44622853-f6881e80-a886-11e8-834a-b3e67dd3c925.PNG)
![entry point not found 4](https://user-images.githubusercontent.com/2982395/44622854-f7b94b80-a886-11e8-9c2c-7624812de972.PNG)
Missing Entry Points

![project directory](https://user-images.githubusercontent.com/2982395/44622858-09025800-a887-11e8-8dd9-a38bbac90341.PNG)
Project Directory

![toolchain executables](https://user-images.githubusercontent.com/2982395/44622859-10296600-a887-11e8-904c-da5c17ae46d6.PNG)
![search directories - compiler](https://user-images.githubusercontent.com/2982395/44622860-115a9300-a887-11e8-9a17-220c1794415b.PNG)
![search directories - linker](https://user-images.githubusercontent.com/2982395/44622861-128bc000-a887-11e8-90fd-66cf1823bc6d.PNG)
Compiler settings

![sfml 2 3 2](https://user-images.githubusercontent.com/2982395/44622866-1fa8af00-a887-11e8-8ac0-2da17b89ad98.PNG)
![codeblocks16mingw](https://user-images.githubusercontent.com/2982395/44622867-20d9dc00-a887-11e8-9597-b06f356a1678.PNG)
SFML 2.3.2 and CodeBlocks 32 bit Version 16.01

Do I have the wrong DLLs included?

Should I update to Code:Blocks 17.12 and get SFML 2.5.0 [GCC 5.1.0 TDM (SJLJ) - Code::Blocks - 32-bit] and drmingwa 0.8.2? I was trying to match the versions in the walkthrough to replicate the instructions before trying something else.

Or should I try a cmake approach to building? Migh tbe simpler, but the WIKI doesn't fully describe that approach.

EDIT: Also tried to build Release and same issues happened

**Comment by msbeck** [comment]
32-bit. I tried to stick to 32bit on all downloads.

**Comment by NinjaSpectre** [comment]
You have the wrong SFML DLLs.  I believe the the compiler in that Code::Blocks download is GCC 4.9.2 TDM. Running `gcc -v` from the windows command line would confirm what version you have.

I have used that compiler version with SFML 2.4.2 and drmingw 0.8.2 to build a Windows release at master.  You should only need to edit EmptyEpsilon.cbp so that the link directory paths are correct for your setup.

I tried using a cmake setup but it doesn't work so great on Windows so my advice would be to avoid it. Of course YMMV :)

**Comment by JeffM2501** [comment]
I was able to get it building on windows with Code::Blocks.

The current installer for Code::Blocks is using GCC 5.1.0 TDM, thereforee you must get the SFML 2.5.0 that is built for it. GCC 5.1.0 TDM (SJLJ) - Code::Blocks - 32-bit from https://www.sfml-dev.org/download/sfml/2.5.0/

The SFML and compiler versions must match identically because of the unique way that MinGW does dynamic linking on windows (in the least dynamic way possible)

With that version it will build. After you get through the torrent of warnings, you'll get a .exe.

Move that exe into the root of the EmptyEpsilon dir, and grab the dlls from SFML, drmingw-0.8.2-win32, and the NinGW that ships with Code::Blocks (C:\Program Files (x86)\CodeBlocks\MinGW\bin). 

I had to copy some of the oog and vorbis libs from the bin install to get everything working, so if you see any dlls in the zip file that you can't find on your system, just copy them over (libJpeg, libVorbis, etc..)

After that it seems to run ok.

Code::Blocks has some serious stability and usability problems, so I don't know how useful it is for actual development, but it can be used to at least spit out a binary.

**Comment by msbeck** [comment]
@ninjaspectre The SFML I had in my screenshot was for GCC 4.9.2. What's the TDM part?

I'll try the latest versions as Jeff suggests and see what I get.

**Comment by NinjaSpectre** [comment]
> What's the TDM part?

A _customized_ MinGW and MinGW-w64 hybrid compiler which can create 32-bit or 64-bit binaries.

The DLLs you need to link with also must have been made with this TDM compiler, the ones provided for SFML 2.3.2 are for GCC versions 4.7.1 and 4.8.1

**Comment by msbeck** [comment]
The arrow I have is pointing at GCC 4.9.2. MinGW (DWM) - 32 bit. Are you saying I need to find a GCC 4.9.2. MinGW (TDM) - 32 bit somewhere?

And are the DLLs not the ones in the SFML folders like the wiki suggests? That's where I grabbed them all.

**Comment by NinjaSpectre** [comment]
Right, provided that you are using the GCC 4.9.2 TDM compiler (I am pretty sure that is what the problem is). There are GCC 4.9.2 TDM (SJLJ) - 32 bit DLLs provided for [SMFL 2.4.2](https://www.sfml-dev.org/download/sfml/2.4.2/).

The wiki is a little outdated :)

**Comment by msbeck** [comment]
Ok, I'm confused by the acronyms. Are you saying mine is wrong because its DWM instead of TDM?

Thanks!

**Comment by NinjaSpectre** [comment]
If you are not familiar with the naming conventions it can be extremely confusing/frustrating.
What I am trying to say is that the SFML DLLs are made with different types of GCC compilers... TDM and MinGW... and without going into loads of detail the differences between the two are enough to make them not work with each other.
As far as what is in the parentheses that is another ball of wax.  That tells the method used for handling exceptions, and there are different types available here as well.

**Comment by oznogon** [comment]
You must use the SFML libraries that match the compiler. There are two relevant compilers for Windows: TDM and MinGW. Code::Blocks ships with TDM, so you have to use the SFML libraries for TDM.

The DWM, DW2, SJLJ parentheticals aren't directly relevant to this point — that's the exception handling implementation, not the compiler.

![screen shot 2018-08-27 at 10 48 54 am](https://user-images.githubusercontent.com/19192104/44676049-00289800-a9e7-11e8-86ea-daead711031b.png)

**Comment by oznogon** [comment]
From the Code::Blocks site, right beneath that screenshot, emphasis theirs:

> **NOTE:** The codeblocks-17.12mingw-setup.exe file includes additionally the GCC/G++ compiler and GDB debugger from **TDM-GCC** (version 5.1.0, 32 bit, SJLJ).

SFML tutorials:

> If you are unsure, check which of the libgcc_s_sjlj-1.dll or libgcc_s_dw2-1.dll files is present in your MinGW/bin folder. If MinGW was installed along with Code::Blocks, you probably have an SJLJ version. 

Looking at the no-install Code::Blocks 17.12 archive, both SJLJ and DW2 libraries might be included. If installed through the executable installer, it might depend on the installation options.

To @msbeck's pertinent point, if the MinGW/DW2 SFML libraries work, great. If they don't, try the other TDM/SJLJ set of SFML libraries instead.

**Comment by daid** [comment]
Yes, this is super annoying, it needs to match 100% or you get the above problems.

I'm using code::blocks without mingw, and use the DW2 mingw64.


Alternative is to compile your own version of SFML, with CMake. The CMake scripts also work on windows.

(Side note, the releases on the website are cross compiled, so build on Linux, to be ran on Windows. This also contains a custom compiled SFML. So don't use those DLLs. As you will run into compiler compatibility issues)

**Comment by msbeck** [comment]
I have matched the two.

<img width="744" alt="screen shot 2018-08-27 at 5 23 16 pm" src="https://user-images.githubusercontent.com/2982395/44689664-ef474900-aa1d-11e8-999d-f6708363e2f3.png">

I used Codeblocks 16.01 MinGW with the following specs:

> The installers/packages with compiler include the GNU compiler suite
> and GNU debugger (GDB) from TDM-GCC (version 4.9.2, 32 bit, SJLJ).

So I used the "GCC 4.9.2. MinGW (DWM) - 32 bit" version of SFML 2.3.2. 

But all this doesn't matter if I can get it running using the latest like Jeff did - provided he played the game and it didn't crash and so forth.

**Comment by JeffM2501** [comment]
I started a server, connected a helm client to it and flew around. I would not call that extensive testing, but it seemed to function at some level.

**Comment by JeffM2501** [comment]
I did another CodeBlocks install and build on a clean machine yesterday with out issue, following the same simple steps. So it wasn't a fluke that I was able to build.

Steps:
Get Code::Blocks, get SFML for code blocks, get drmingw, edit EmptyEpsilon.cbp to point to SFML 2.5.0 and drmingw 0.8.2, build, wait, copy dlls, run..
It even works with a portable install of codeblocks.

![image](https://user-images.githubusercontent.com/322174/44805876-3ba79b80-ab7a-11e8-87b3-d5c6cd744016.png)

I did tell code::blocks to use the "gcc" compiler (the first one in the list) when it asked.I'm not sure if that's MinGW or TDM, my guess is it's TDM because it's working with the TDM SFML.

**Comment by JordanLongstaff** [comment]
I'm trying to use Code::Blocks to build EmptyEpsilon and so far I can't even build it because I am getting tons of errors that all say things like "undefined reference to `_Unwind_Resume'" and "undefined reference to `__gxx_personality_v0'". I'm using Code::Blocks 17.12 with MinGW GCC 5.1.0 included, along with SFML 2.5.0 (GCC 5.1.0 TDM (SJLJ) - Code::Blocks) and drmingw 0.8.2.

**Comment by daid** [comment]
@JordanLongstaff Those errors mean the gcc version you use and a library (most likely SFML) don't match up.

**Comment by JordanLongstaff** [comment]
@daid They should, according to several of the previous comments in this thread. I don't know where the mismatch is.

**Comment by daid** [comment]
Slightly more details, both those symbols are part of the C++ exception handling of GCC. The exact details are set during compile time, and thus are set when SFML was compiled. And then later for

**Comment by JordanLongstaff** [comment]
OK, after a bit of mucking about with my compiler settings, I finally managed to build it and run it. 😃

I see a couple of extra buttons in the main menu ("Models!" and "To Da GM!"). I assume these are there because of the Debug build. I tried switching to Release and now the compiler can't seem to find SeriousProton/src/Box2D/Collision/b2Collision.h.

**Comment by JordanLongstaff** [comment]
And now my build crashes every time I try to start a scenario in a server. 💢

**Comment by amir-arad** [comment]
I just followed the instructions on the [wiki](https://github.com/daid/EmptyEpsilon/wiki/Build-from-sources#windows-codeblocks-and-mingw) with the helpful pointers laid out here and managed to compile a windows version within a few hours, after spending nights trying to cross-compile to windows from linux.
Thanks people.

**Comment by daid** [comment]
Yes, the debug build shows a few extra options in the main menu (so you can test things faster). It also shows textures that it cannot find as a black/pink checkerboard.

**Comment by JordanLongstaff** [comment]
What is this "sym" folder that appears inside the build folder?

**Comment by JordanLongstaff** [comment]
Do we really need to add the drmingw executable into the build folder? The releases don't include it.

**Comment by daid** [comment]
Part of SFML.

**Comment by JordanLongstaff** [comment]
Really? Because I can't find them anywhere in the SFML files and the wiki does not mention them.

**Comment by daid** [comment]
https://github.com/SFML/SFML/tree/master/extlibs/libs-mingw/x86

Seems like the default SFML version links those static, while I linked them dynamic in my distribution.

**Comment by JordanLongstaff** [comment]
So that's the difference? If so, then I guess it doesn't matter if I don't include those DLLs. I'm just trying to dress up my custom Release build to look official.

Can you answer my other questions about the "sym" folder and the drmingw EXE?

**Comment by daid** [comment]
I'm not getting a sym folder, what kind of files are in there?
The drmingw exe isn't needed, only the others, if you want to include the crash log generator.

**Comment by JordanLongstaff** [comment]
I attached a screenshot of the one folder inside the sym folder. Unfortunately I can't open it up further because I already wiped the Release build containing it along with the backup.

**Comment by JordanLongstaff** [comment]
Oh wait, you know what I think it was? I think it was created by my attempting to debug using Code::Blocks. I think I did that while I was debugging my "auto-reloading weapon tubes" feature.
